patroni/postgresql/sync.py

```diff
@@ -239,52 +232,119 @@ def current_state(self, cluster: Cluster) -> Tuple[CaseInsensitiveSet, CaseInsen
         for pid, app_name, sync_state, replica_lsn in sorted(pg_stat_replication, key=lambda r: r[2:4], reverse=True):
             member = members.get(app_name)
             if member and member.is_running and not member.tags.get('nosync', False):
-                replica_list.append((pid, member.name, sync_state, replica_lsn, bool(member.nofailover)))
+                yield (pid, member.name, sync_state, replica_lsn, bool(member.nofailover))
 
-        max_lsn = max(replica_list, key=lambda x: x[3])[3]\
-            if len(replica_list) > 1 else self._postgresql.last_operation()
+    def _process_replica_readiness(self, cluster: Cluster, replica_list: List[Tuple[int, str, str, int, bool]]) -> None:
+        """Flags replicas as truely "synchronous" when they caught up with "_primary_flush_lsn"."""
+        if TYPE_CHECKING:  # pragma: no cover
+            assert self._postgresql.global_config is not None
+        for pid, app_name, sync_state, replica_lsn, _ in replica_list:
+            if app_name not in self._ready_replicas and app_name in self._ssn_data.members:
+                if self._postgresql.global_config.is_quorum_commit_mode:
+                    # When quorum commit is enabled we can't check against cluster.sync because nodes
+                    # are written there when at least one of them caught up with _primary_flush_lsn.
+                    if replica_lsn >= self._primary_flush_lsn\
+                            and (sync_state == 'quorum' or (not self._postgresql.supports_quorum_commit
+                                                            and sync_state in ('sync', 'potential'))):
+                        self._ready_replicas[app_name] = pid
+                elif cluster.sync.matches(app_name) or sync_state == 'sync' and replica_lsn >= self._primary_flush_lsn:
+                    # if standby name is listed in the /sync key we can count it as synchronous, otherwise it becomes
+                    # "really" synchronous when sync_state = 'sync' and we known that it managed to catch up
+                    self._ready_replicas[app_name] = pid
+
+    def current_state(self, cluster: Cluster) -> _SyncState:
+        """Finds best candidates to be the synchronous standbys.
+
+        Current synchronous standby is always preferred, unless it has disconnected or does not want to be a
+        synchronous standby any longer.
+
+        Standbys are selected based on values from the global configuration:
```

Comment:
The bullet points need to be indented and have line breaks after each one to be properly formatted in sphinx